File: API/1084_MTF_Oscillator_Framework/CS/MtfOscillatorFrameworkStrategy.cs
using System;
using System.Linq;

using Ecng.Common;

using StockSharp.Algo;
using StockSharp.Algo.Indicators;
using StockSharp.Algo.Strategies;
using StockSharp.BusinessEntities;
using StockSharp.Messages;

namespace StockSharp.Samples.Strategies;

/// <summary>
/// Multi-timeframe RSI strategy.
/// </summary>
public class MtfOscillatorFrameworkStrategy : Strategy
{
	private readonly StrategyParam<DataType> _candleType;
	private readonly StrategyParam<int> _rsiLength;
	private readonly StrategyParam<decimal> _overbought;
	private readonly StrategyParam<decimal> _oversold;
	private readonly StrategyParam<int> _cooldownBars;

	private decimal _prevRsi;
	private bool _hasPrev;
	private int _barIndex;
	private int _lastSignalBar = -1000000;

	public DataType CandleType { get => _candleType.Value; set => _candleType.Value = value; }
	public int RsiLength { get => _rsiLength.Value; set => _rsiLength.Value = value; }
	public decimal Overbought { get => _overbought.Value; set => _overbought.Value = value; }
	public decimal Oversold { get => _oversold.Value; set => _oversold.Value = value; }
	public int CooldownBars { get => _cooldownBars.Value; set => _cooldownBars.Value = value; }

	public MtfOscillatorFrameworkStrategy()
	{
		_candleType = Param(nameof(CandleType), TimeSpan.FromMinutes(5).TimeFrame());
		_rsiLength = Param(nameof(RsiLength), 14);
		_overbought = Param(nameof(Overbought), 68m);
		_oversold = Param(nameof(Oversold), 32m);
		_cooldownBars = Param(nameof(CooldownBars), 6).SetGreaterThanZero();
	}

	protected override void OnStarted2(DateTime time)
	{
		base.OnStarted2(time);
		_prevRsi = 0m;
		_hasPrev = false;
		_barIndex = 0;
		_lastSignalBar = -1000000;

		var rsi = new RelativeStrengthIndex { Length = RsiLength };
		var ema = new ExponentialMovingAverage { Length = 20 };

		var subscription = SubscribeCandles(CandleType);
		subscription
			.Bind(rsi, ema, ProcessCandle)
			.Start();
	}

	private void ProcessCandle(ICandleMessage candle, decimal rsi, decimal ema)
	{
		if (candle.State != CandleStates.Finished)
			return;

		_barIndex++;

		if (!_hasPrev)
		{
			_prevRsi = rsi;
			_hasPrev = true;
			return;
		}

		var canSignal = _barIndex - _lastSignalBar >= CooldownBars;
		var longSignal = _prevRsi <= Oversold && rsi > Oversold;
		var shortSignal = _prevRsi >= Overbought && rsi < Overbought;

		if (canSignal && longSignal && Position <= 0)
		{
			BuyMarket();
			_lastSignalBar = _barIndex;
		}
		else if (canSignal && shortSignal && Position >= 0)
		{
			SellMarket();
			_lastSignalBar = _barIndex;
		}

		if (Position > 0 && rsi >= 60m)
			SellMarket();
		else if (Position < 0 && rsi <= 40m)
			BuyMarket();

		_prevRsi = rsi;
	}

	/// <inheritdoc />
	protected override void OnReseted()
	{
		base.OnReseted();

		_prevRsi = 0m;
		_hasPrev = false;
		_barIndex = 0;
		_lastSignalBar = -1000000;
	}
}

File: API/1001-1100/1084_MTF_Oscillator_Framework/PY/mtf_oscillator_framework_strategy.py
import clr

clr.AddReference("StockSharp.Messages")
clr.AddReference("StockSharp.Algo")
clr.AddReference("StockSharp.Algo.Indicators")
clr.AddReference("StockSharp.Algo.Strategies")

from System import TimeSpan
from StockSharp.Messages import DataType, CandleStates
from StockSharp.Algo.Indicators import RelativeStrengthIndex, ExponentialMovingAverage
from StockSharp.Algo.Strategies import Strategy

class mtf_oscillator_framework_strategy(Strategy):
    """
    MTF Oscillator Framework: RSI zone crossover with exits at neutral levels.
    """

    def __init__(self):
        super(mtf_oscillator_framework_strategy, self).__init__()
        self._rsi_length = self.Param("RsiLength", 14).SetDisplay("RSI", "RSI period", "Indicators")
        self._overbought = self.Param("Overbought", 68.0).SetDisplay("OB", "Overbought", "Signals")
        self._oversold = self.Param("Oversold", 32.0).SetDisplay("OS", "Oversold", "Signals")
        self._cooldown_bars = self.Param("CooldownBars", 6).SetDisplay("Cooldown", "Min bars between entries", "Risk")
        self._candle_type = self.Param("CandleType", DataType.TimeFrame(TimeSpan.FromMinutes(5))).SetDisplay("Candle Type", "Candles", "General")

        self._prev_rsi = 0.0
        self._has_prev = False
        self._bar_index = 0
        self._last_signal_bar = -1000000

    @property
    def candle_type(self):
        return self._candle_type.Value

    def OnReseted(self):
        super(mtf_oscillator_framework_strategy, self).OnReseted()
        self._prev_rsi = 0.0
        self._has_prev = False
        self._bar_index = 0
        self._last_signal_bar = -1000000

    def OnStarted2(self, time):
        super(mtf_oscillator_framework_strategy, self).OnStarted2(time)
        rsi = RelativeStrengthIndex()
        rsi.Length = self._rsi_length.Value
        ema = ExponentialMovingAverage()
        ema.Length = 20
        subscription = self.SubscribeCandles(self.candle_type)
        subscription.Bind(rsi, ema, self._process_candle).Start()
        area = self.CreateChartArea()
        if area is not None:
            self.DrawCandles(area, subscription)
            self.DrawIndicator(area, rsi)
            self.DrawOwnTrades(area)

    def _process_candle(self, candle, rsi_val, ema_val):
        if candle.State != CandleStates.Finished:
            return
        rsi = float(rsi_val)
        self._bar_index += 1
        if not self._has_prev:
            self._prev_rsi = rsi
            self._has_prev = True
            return
        ob = float(self._overbought.Value)
        os_val = float(self._oversold.Value)
        can_signal = self._bar_index - self._last_signal_bar >= self._cooldown_bars.Value
        long_signal = self._prev_rsi <= os_val and rsi > os_val
        short_signal = self._prev_rsi >= ob and rsi < ob
        if can_signal and long_signal and self.Position <= 0:
            self.BuyMarket()
            self._last_signal_bar = self._bar_index
        elif can_signal and short_signal and self.Position >= 0:
            self.SellMarket()
            self._last_signal_bar = self._bar_index
        if self.Position > 0 and rsi >= 60.0:
            self.SellMarket()
        elif self.Position < 0 and rsi <= 40.0:
            self.BuyMarket()
        self._prev_rsi = rsi

    def CreateClone(self):
        return mtf_oscillator_framework_strategy()
